04-onion-routing.md

```diff
@@ -1407,6 +1438,117 @@ The _origin node_:
+locked in the channel.
+
+For every resolved incoming HLTC a peer has forwarded through a node, its 
+`effective_fees` are calculated as follows:
+- if `endorsed` = 1 in `update_add_htlc`:
+  - `effective_fees` = `fees` - `opportunity_cost` 
+- otherwise: 
+  - if successfully resolved AND `resolution_time` <= `resolution_period`
+    - `effective_fees` = `fees` * 0.5
+  - otherwise: 
+    - `effective_fees` = 0
+
+Incoming in-flight HTLCs have a negative impact on reputation, as their 
+influence is unknown until time of resolution. The `outstanding_risk` of each 
+in flight HTLC is calculated as follows: 
+- if `endorsed` = 1 in `update_add_htlc`:
+  - `outstanding_risk` = `fees` * ( `cltv_expiry` - `height_added` * 10 * 60), 
+    where `height_added` is the block height at which the HLTC was irrevocably
+    committed to by the receiving node.
+
+The `effective_fees` that a peer has paid our node are compared to our total 
+routing revenue to classify a peer's reputation as "good" or "neutral". This
+relates the fees that must be paid to earn "good" reputation to the damage that 
+can be done by abusing it.
+
+Routing revenue is assessed over a sliding window, so that reputation 
+classification is related to the node's current routing activity. Nodes MAY
+choose a window per their preferences - shorter windows being more reactive to 
+recent routing patterns, and longer windows aggregating trends over time.
+
+- `routing_window`: the period of time in which total routing revenue is 
+  assessed (default: 10 minutes * the maximum number of blocks a HTLC may be
+  held before the node will send `expiry_too_far`, as outlined in [Failure Messages](#failure-messages)).
+- `routing_revenue`: the sum of all fees paid to the node to forward HTLCs
+  over the interval [ `now` - `routing_window` ; `now` ].
+
+The total `effective_fees` that an individual peer has generated are assessed
+over a longer period of time to relate its reputation classification to the 
+routing activity of the node (represented by `routing_revenue`).
+
+- `reputation_window`: the period of time over which the node's `effective_fees`
+  are calculated to classify its reputation (recommended default = 10 * `routing_window`).
+- `reputation_revenue`: the total `effective_fees` of HTLCs forwarded by the peer
+  over the interval [ `now` - `reputation_window` ; `now` ] less the sum of 
+  `outstanding_risk` for all incoming, in-flight HLTCs from the peer.
+
+A peer is classified as having "good" local reputation iff `reputation_revenue` >= `routing_revenue`, 
+otherwise the peer's reputation is classified as "neutral".
+
+### Rationale
+
+A jamming attack can only be achieved with good reputation, as the resources 
+available to neutral reputation nodes are limited by resource bucketing. If a 
+node is attacked, it is guaranteed to have been paid at least its 
+`routing_revenue` over the previous `reputation_window`.
```

Comment:
I think `routing_revenue` should be abstracted and rather say `reputation acquisition cost` over the previous `reputation_windows`. `reputation acquisition cost` is looser than pure fee-denominated revenue and allows a routing node to encompass other decision factors (e.g if the HTLC coming from a “well-known” LSP), while maintaining the same resource control equation.

Reply:
This solution looks at reputation based on fees. LSPs will have their own custom logic.

Reply:
Well ideally LSPs and any other entity can build their own custom logic based on the reputation algorithm interface.